storey/writers.py

```diff
@@ -652,12 +652,23 @@ def __init__(self, table: Union[Table, str], columns: Optional[List[str]] = None
         self._write_missing_fields = False
 
     def _init(self):
-        _ConcurrentByKeyJobExecution._init(self)
         _Writer._init(self)
 
-    async def _process_events(self, events):
-        data_to_persist = self._event_to_writer_entry(events[-1])
-        return await self._table._persist_key(events[0].key, data_to_persist)
-
     async def _handle_completed(self, event, response):
         await self._do_downstream(event)
+
+    async def _do(self, event):
+        if event is _termination_obj:
+            await self._table._terminate()
+            return await self._do_downstream(_termination_obj)
+
+        if event.key is None:
+            raise ValueError("Can't write to table without a key")
```

Comment:
Could be clearer, e.g. "Event could not be written to table because it has no key"